blockmanager.go

```diff
@@ -34,48 +37,100 @@ const (
 	blockDbNamePrefix = "blocks"
 )
 
+// zeroHash is the zero value hash (all zeros).  It is defined as a convenience.
+var zeroHash wire.ShaHash
+
 // newPeerMsg signifies a newly connected peer to the block handler.
 type newPeerMsg struct {
-	peer *peer
+	peer *peer.Peer
+}
+
+// peerInfo extends peer with data required by the blockmanager.
+type peerInfo struct {
+	*peer.Peer
+
+	continueHash    *wire.ShaHash
+	relayMtx        sync.Mutex
+	disableRelayTx  bool
+	requestQueue    []*wire.InvVect
+	requestedTxns   map[wire.ShaHash]struct{}
+	requestedBlocks map[wire.ShaHash]struct{}
+	filter          *bloom.Filter
+	knownAddrMutex  sync.Mutex
+	knownAddresses  map[string]struct{}
+
+	// The following chans are used to sync blockmanager and server.
+	txProcessed    chan struct{}
+	blockProcessed chan struct{}
+}
+
+// relayTxDisabled returns whether or not relaying of transactions for the given
+// peer is disabled.
+// It is safe for concurrent access.
+func (p *peerInfo) relayTxDisabled() bool {
+	p.relayMtx.Lock()
+	defer p.relayMtx.Unlock()
+
+	return p.disableRelayTx
+}
+
+// addKnownAddress adds the given address to the set of addresses known to the
+// peer.
+// It is safe for concurrent access.
+func (p *peerInfo) addKnownAddress(addr string) {
+	p.knownAddrMutex.Lock()
+	defer p.knownAddrMutex.Unlock()
```

Comment:
I would avoid the defer here since there's no complex branching that the defer helps to make readable and I don't trust the compiler to optimize this by avoiding the defer overhead.